File: solutions/443_compress.ts
/**
 *
 * @param chars
 * @returns
 * Time: O(n)
 * Space: O(1)
 */
function compress(chars: string[]): number {
  let read = 0;
  let write = 0;

  while (read < chars.length) {
    let count = 0;
    let currentChar = chars[read];
    while (read < chars.length && chars[read] === currentChar) {
      read++;
      count++;
    }
    chars[write++] = currentChar;
    if (count > 1) {
      for (let c of count.toString().split("")) {
        chars[write++] = c;
      }
    }
  }

  chars.length = write;
  return write;
}

File: solutions_py/443_compress.py
class Solution(object):
    def compress(self, chars):
        """
        :type chars: List[str]
        :rtype: int
        """

        read, write = 0, 0
        while read < len(chars):
            chars[write] = chars[read]
            count = 1

            while read + 1 < len(chars) and chars[read] == chars[read + 1]:
                read += 1
                count += 1

            if count > 1:
                for c in str(count):
                    chars[write + 1] = c
                    write += 1

            read += 1
            write += 1

        return write
